src/hist/core.py

```diff
@@ -45,24 +45,33 @@ def __call__(self, *args, **kwargs):
 
 @hist.utils.set_family(hist.utils.HIST_FAMILY)
 class BaseHist(bh.Histogram):
+    __slots__ = ("_ax", "_storage_proxy")
+
     def __init__(self, *args, **kwargs):
         """
             Initialize BaseHist object. Axis params can contain the names.
         """
+        self._ax: List[Any] = []
+        self._hist: Any = None
+        self._storage_proxy: Any = None
+
         if len(args):
-            super().__init__(*args, **kwargs)
-            self.names: dict = dict()
-            for ax in self.axes:
-                if ax.name and ax.name in self.names:
-                    raise KeyError(
-                        f"{self.__class__.__name__} instance cannot contain axes with duplicated names."
-                    )
-                else:
-                    self.names[ax.name] = True
-        else:
-            self._ax = []
             self._hist = None
-            self._storage_proxy = None
+            self._ax = []
+            super().__init__(*args, **kwargs)
+            valid_names = [ax.name for ax in self.axes if ax.name]
+            if len(valid_names) != len(set(valid_names)):
+                raise KeyError(
+                    f"{self.__class__.__name__} instance cannot contain axes with duplicated names"
+                )
+            for i, ax in enumerate(self.axes):
+                # title will return name if title is not set, so this is safe
+                if not ax.title:
```

Comment:
why not

```python
ax.title = ax.title or ax.name or f"Axis {i}"
```

Reply:
Because if `metadata["title"]` is blank, then `.title` returns `metadata["name"]`, so if `.title` is blank, `.name` must be blank.

Reply:
Your way would work if we want to statically set the title always, and leave the logic out of the getters for `.title`. I _think_ I prefer the current method, because it doesn't add a title when there is a name, it just returns the name as the title when asked.

Reply:
We need to have a test to make sure adding the same axes to two Histograms property makes a copy, rather than causing a battle as to who gets to set the title.

Reply:
https://github.com/scikit-hep/hist/blob/7e76c3f0dd7b4014d5a32f1c048d200bc204eb77/src/hist/core.py#L68

Sorry, I skipped this line 😄

Reply:
> We need to have a test to make sure adding the same axes to two Histograms property makes a copy, rather than causing a battle as to who gets to set the title.

Better open in a new issue. Let me do it!